drivers/counter/counter_mchp_tcc_g1.c

```diff
@@ -0,0 +1,784 @@
+	struct tcc_counter_irq_map *channel_irq_map;
+	uint32_t max_bit_width;
+	uint16_t prescaler;
+	uint8_t max_channels;
+	void (*irq_config_func)(const struct device *dev);
+};
+
+typedef enum {
+	/* Reload or reset the counter on the next generic clock */
+	TCC_GCLK_RESET_ON_GENERIC_CLOCK = 0x0,
+	/* Reload or reset the counter on the next prescaler clock */
+	TCC_PRESC_RESET_ON_PRESCALER_CLOCK = 0x1,
+	/* Reload or reset the counter on the next generic clock, and reset the prescaler
+	 * counter
+	 */
+	TCC_RESYNC_RESET_ON_GENERIC_CLOCK = 0x2,
+} tcc_counter_prescaler_sync_mode_t;
+
+/* find the prescale index based on the value present in the device tree */
+static uint8_t get_tcc_prescale_index(uint16_t prescaler)
+{
+	uint8_t prescale_index = 0u;
+
+	switch (prescaler) {
+	case 64:
+		prescale_index = TCC_CTRLA_PRESCALER_DIV64_Val;
+		break;
+	case 256:
+		prescale_index = TCC_CTRLA_PRESCALER_DIV256_Val;
+		break;
+	case 1024:
+		prescale_index = TCC_CTRLA_PRESCALER_DIV1024_Val;
+		break;
+	default:
+		if ((prescaler >= 1) && (prescaler <= 16)) {
+			prescale_index = __builtin_ctz(prescaler);
+		}
+		break;
+	}
+
+	return prescale_index;
+}
+
+static void tcc_counter_wait_sync(const volatile uint32_t *sync_reg_addr, uint32_t bit_mask)
+{
+	if (false == WAIT_FOR((((*sync_reg_addr) & bit_mask) == 0u),
+			      TCC_SYNCHRONIZATION_TIMEOUT_IN_US, k_busy_wait(DELAY_US))) {
+		LOG_ERR("%s : Synchronization time-out occurred", __func__);
+	}
+}
+
+static void tcc_counter_ctrlbset_sync(tcc_registers_t *const p_regs, uint32_t bit_mask)
+{
+	if (false == WAIT_FOR((((p_regs->TCC_CTRLBSET) & bit_mask) == 0u), TCC_CTRLB_TIMEOUT_IN_US,
+			      k_busy_wait(DELAY_US))) {
+		LOG_ERR("%s : CTRLBSET time-out occurred", __func__);
+	}
+}
+
+static void tcc_counter_init(tcc_registers_t *const p_regs, uint32_t prescaler,
+			     uint8_t max_channels, const uint32_t max_bit_width)
+{
+	uint32_t max_counter_value;
+
+	p_regs->TCC_CTRLA |= TCC_CTRLA_SWRST_Msk;
+	tcc_counter_wait_sync(&p_regs->TCC_SYNCBUSY, TCC_SYNCBUSY_SWRST_Msk);
+
+	/* Configure counter mode & prescaler */
+	p_regs->TCC_CTRLA = TCC_CTRLA_CPTEN0(0U) | TCC_CTRLA_CPTEN1(0U) | TCC_CTRLA_CPTEN2(0U) |
+			    TCC_CTRLA_CPTEN3(0U) | TCC_CTRLA_CPTEN4(0U) | TCC_CTRLA_CPTEN5(0U) |
+			    TCC_CTRLA_MSYNC(0U) |
+			    TCC_CTRLA_PRESCALER(get_tcc_prescale_index(prescaler)) |
+			    TCC_CTRLA_PRESCSYNC(TCC_GCLK_RESET_ON_GENERIC_CLOCK) |
+			    TCC_CTRLA_RUNSTDBY(0U) | TCC_CTRLA_RESOLUTION(0U);
+
+	/* Configure waveform generation mode */
+	p_regs->TCC_WAVE = TCC_WAVE_WAVEGEN_NFRQ;
+
+	/* Configure timer one shot mode & direction */
+	p_regs->TCC_CTRLBSET = TCC_CTRLBCLR_ONESHOT(0U) | TCC_CTRLBCLR_DIR(0U);
+
+	/* Configure drive control register */
+	p_regs->TCC_DRVCTRL = TCC_DRVCTRL_INVEN(0U);
+
+	/* Set the period register to maximum value */
+	max_counter_value = (uint32_t)(BIT64(max_bit_width) - 1);
+
+	p_regs->TCC_PER = max_counter_value;
+
+	for (uint8_t index = 0u; index < max_channels; index++) {
```

Comment:
Should we capture a error related to the `max_channels `? I mean what is the maximum channels available.